### PR TITLE
Add hook to restrict fields and fix field renames in fragments

### PR DESCRIPTION
**Pull request by andimarek**
adds the ability to filter fields which are not send to an underlying service and added as `null` value for the overall response.

It also contains a fix for Field renames in Fragments.

### CODE REVIEW

File: src/main/java/graphql/nadel/engine/Execution.java [hunk 1]
```diff
@@ -47,12 +50,29 @@
...
+                     IntrospectionRunner introspectionRunner,
+                     ServiceExecutionHooks serviceExecutionHooks,
+                     Object userSuppliedContext) {
         this.services = services;
         this.overallSchema = overallSchema;
         this.instrumentation = instrumentation;
         this.introspectionRunner = introspectionRunner;
-        this.nadelExecutionStrategy = new NadelExecutionStrategy(services, createFieldsInfos(), overallSchema, instrumentation, serviceExecutionHooks);
+        FieldInfos fieldsInfos = createFieldsInfos();
+        if (userSuppliedContext instanceof BenchmarkContext) {
```
Comment (bbakerman) [inline src/main/java/graphql/nadel/engine/Execution.java:66]: Whats is this for?  Profilable code??

File: src/main/java/graphql/nadel/engine/ExecutionResultNodeMapper.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+//            return assertNotNull(outputType, "type %s not found in overall schema for field type", ((GraphQLNamedOutputType) graphQLOutputType).getName());
+//        }
+//        if (isList(graphQLOutputType)) {
+//            return list(mapFieldType((GraphQLOutputType) ((GraphQLList) graphQLOutputType).getWrappedType(), overallSchema, typeRenameMappings));
+//        }
+//        if (isNonNull(graphQLOutputType)) {
+//            return nonNull(mapFieldType((GraphQLOutputType) ((GraphQLNonNull) graphQLOutputType).getWrappedType(), overallSchema, typeRenameMappings));
+//        }
+//        return Assert.assertShouldNeverHappen();
+//    }
```
Comment (bbakerman) [inline src/main/java/graphql/nadel/engine/ExecutionResultNodeMapper.java:73]: remove dead code or leave it uncalled - but dont leave it commented out

File: src/main/java/graphql/nadel/engine/FixListNamesAdapter.java [hunk 3]
```diff
@@ -38,7 +38,8 @@ public ExecutionResultNode withNewChildren(ExecutionResultNode node, Map<String,
     }
 
     private ExecutionResultNode fixFieldName(ListExecutionResultNode node, ExecutionResultNode childNode) {
-        return StrategyUtil.changeFieldInResultNode(node, childNode.getMergedField());
+//        return StrategyUtil.changeFieldInResultNode(node, childNode.getMergedField());
```
Comment (bbakerman) [inline src/main/java/graphql/nadel/engine/FixListNamesAdapter.java:41]: clean up comment

File: src/main/java/graphql/nadel/engine/HydrationInputResolver.java [hunk 4]
```diff
@@ -242,27 +237,35 @@ private boolean isBatchHydrationField(HydrationInputNode hydrationInputNode) {
                         topLevelFieldType,
                         serviceExecutionHooks,
                         service,
-                        serviceContexts.get(service));
+                        serviceContexts.get(service)
+                );
 
 
-        fieldTracking.fieldsDispatched(singletonList(hydratedFieldStepInfo));
+//        fieldTracking.fieldsDispatched(singletonList(hydratedFieldStepInfo));
```
Comment (bbakerman) [inline src/main/java/graphql/nadel/engine/HydrationInputResolver.java:244]: clean up dead code

File: src/main/java/graphql/nadel/engine/HydrationInputResolver.java [hunk 5]
```diff
@@ -464,28 +476,28 @@ private ObjectExecutionResultNode findMatchingResolvedNode(ExecutionContext exec
...
+//        if (ExecutionPathUtils.isListEndingPath(path)) {
+//            //
+//            // a path like /issues/comments[0] wont ever have had a /issues/comments path completed but the tracing spec needs
+//            // one so we make one
+//            ExecutionPath newPath = ExecutionPathUtils.removeLastSegment(path);
+//            ExecutionStepInfo newStepInfo = hydratedFieldStepInfo.transform(builder -> builder.path(newPath));
+//            fieldTracking.fieldCompleted(newStepInfo);
+//        }
+//
+//    }
```
Comment (bbakerman) [inline src/main/java/graphql/nadel/engine/HydrationInputResolver.java:500]: clean up dead code

File: src/main/java/graphql/nadel/engine/OverallQueryTransformer.java [hunk 6]
```diff
@@ -430,332 +422,5 @@ private String getUnderlyingTypeNameAndRecordMapping(GraphQLCompositeType typeOv
         return mappingDefinition.getUnderlyingName();
     }
 
-    private class Transformer extends NodeVisitorStub {
```
Comment (bbakerman) [inline src/main/java/graphql/nadel/engine/OverallQueryTransformer.java:433]: Wow thats a LOT less code

File: src/main/java/graphql/nadel/engine/ServiceResultNodesToOverallResult.java [hunk 7]
```diff
@@ -323,54 +425,102 @@ public TraversalControl enter(TraverserContext<ExecutionResultNode> context) {
...
         Set<FieldTransformation> transformations = new LinkedHashSet<>();
-        List<Field> notTransformedFields = new ArrayList<>();
-        for (Field field : mergedField.getFields()) {
-            List<String> fieldIds = FieldMetadataUtil.getRootOfTransformationIds(field);
-            if (fieldIds.size() == 0) {
-                notTransformedFields.add(field);
+        List<String> notTransformedFields = new ArrayList<>();
+        for (String fieldId : node.getFieldIds()) {
+//            System.out.println("processing " + node.getExecutionPath());
+//            System.out.println("field id: " + getId(field) + " field: " + field.getName());
```
Comment (bbakerman) [inline src/main/java/graphql/nadel/engine/ServiceResultNodesToOverallResult.java:457]: clean up comments

File: src/main/java/graphql/nadel/engine/tracking/FieldTracking.java [hunk 8]
```diff
@@ -96,27 +96,26 @@ public void fieldsCompleted(ExecutionResultNode resultNode, Throwable throwable)
...
+//            // we have another call back path for them
+//            if (isHydration(resultNode)) {
+//                continue;
+//            }
+//            completeNode(stepInfo, resultNode, throwable);
+//
+//            // and go down and complete the children
+//            List<ExecutionResultNode> children = resultNode.getChildren();
+//            completeNodes(children, throwable);
+//        }
```
Comment (bbakerman) [inline src/main/java/graphql/nadel/engine/tracking/FieldTracking.java:118]: clean up dead code

File: src/main/java/graphql/nadel/engine/transformation/FieldRenameTransformation.java [hunk 9]
```diff
@@ -60,27 +61,32 @@ public ApplyResult apply(ApplyEnvironment environment) {
...
                                                                 List<FieldTransformation> allTransformations,
+                                                                NormalizedQueryField normalizedQueryField,
                                                                 UnapplyEnvironment environment) {
 
         if (node instanceof ListExecutionResultNode) {
             return mapChildren(node, child -> {
-                ExecutionStepInfo newEsi = replaceFieldsAndTypesWithOriginalValues(allTransformations, child.getExecutionStepInfo(), environment.parentExecutionStepInfo);
-                return replaceFieldsAndTypesInsideList(child.withNewExecutionStepInfo(newEsi),
+//                ExecutionResultNode newChild = replaceFieldWithOriginalValue(allTransformations, child);
+//                newChild = executionResultNodeMapper.mapERNFromUnderlyingToOverall(newChild, environment);
```
Comment (bbakerman) [inline src/main/java/graphql/nadel/engine/transformation/FieldRenameTransformation.java:85]: clean up dead code

File: src/main/java/graphql/nadel/result/ResultNodesCreator.java [hunk 10]
```diff
@@ -1,67 +1,106 @@
...
+//                .fieldIds(fetchedValueAnalysis.getFieldIds())
+//                .objectType(normalizedQueryField.getObjectType())
+//                .fieldDefinition(normalizedQueryField.getFieldDefinition())
+//                .executionPath(fetchedValueAnalysis.getExecutionPath())
+//                .completedValue(createResolvedValue(fetchedValueAnalysis))
+//                .children(childrenNodes)
+//                .build();
+//        return result;
+//    }
+//}
```
Comment (bbakerman) [inline src/main/java/graphql/nadel/result/ResultNodesCreator.java:106]: clean up dead code

File: src/main/java/graphql/nadel/engine/Execution.java (hunk 1 above)
Comment (andimarek) [reply src/main/java/graphql/nadel/engine/Execution.java:66]: yes: this colllects args for the benchmarks

File: src/main/java/graphql/nadel/engine/HydrationInputResolver.java (hunk 4 above)
Comment (andimarek) [reply src/main/java/graphql/nadel/engine/HydrationInputResolver.java:244]: done

File: src/main/java/graphql/nadel/engine/HydrationInputResolver.java (hunk 5 above)
Comment (andimarek) [reply src/main/java/graphql/nadel/engine/HydrationInputResolver.java:500]: done

File: src/main/java/graphql/nadel/engine/ServiceResultNodesToOverallResult.java (hunk 7 above)
Comment (andimarek) [reply src/main/java/graphql/nadel/engine/ServiceResultNodesToOverallResult.java:457]: done

File: src/main/java/graphql/nadel/engine/transformation/FieldRenameTransformation.java (hunk 9 above)
Comment (andimarek) [reply src/main/java/graphql/nadel/engine/transformation/FieldRenameTransformation.java:85]: done

File: src/main/java/graphql/nadel/result/ResultNodesCreator.java (hunk 10 above)
Comment (andimarek) [reply src/main/java/graphql/nadel/result/ResultNodesCreator.java]: done

File: src/main/java/graphql/nadel/engine/tracking/FieldTracking.java (hunk 8 above)
Comment (andimarek) [reply src/main/java/graphql/nadel/engine/tracking/FieldTracking.java]: done

Comment (andimarek) [reply src/main/java/graphql/nadel/engine/tracking/FieldTracking.java]: done

File: src/main/java/graphql/nadel/engine/OverallQueryTransformer.java (hunk 6 above)
Comment (andimarek) [reply src/main/java/graphql/nadel/engine/OverallQueryTransformer.java:433]: it got moved into its own class

File: src/main/java/graphql/nadel/engine/FixListNamesAdapter.java (hunk 3 above)
Comment (andimarek) [reply src/main/java/graphql/nadel/engine/FixListNamesAdapter.java]: done

File: src/main/java/graphql/nadel/engine/ExecutionResultNodeMapper.java (hunk 2 above)
Comment (andimarek) [reply src/main/java/graphql/nadel/engine/ExecutionResultNodeMapper.java]: done

**Pull request merged**